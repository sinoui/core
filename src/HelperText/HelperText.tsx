import styled from 'styled-components';
import getColorFromTheme from '../utils/getColorFromTheme';

interface Props {
  /**
   * 设置为 `true`，表示显示的是校验错误信息。
   */
  error?: boolean;
  /**
   * 设置为 `true`，则表示表单控件处于不可用状态。
   */
  disabled?: boolean;
  /**
   * 指定帮助性文本默认的颜色。默认为 `textSecondary`。
   */
  color?: string;
}

const getColor = ({ error, disabled, color = 'textSecondary' }: Props) => {
  if (error) {
    return 'error';
  }

  if (disabled) {
    return 'textDisabled';
  }

  return color;
};

/**
 * 帮助性文本
 */
const HelperText = styled.div<Props>`
  ${({ theme }) => theme.typography.caption}
  color: ${({ theme, ...rest }) => getColorFromTheme(theme, getColor(rest))};
  text-align: left;
`;

export default HelperText;
