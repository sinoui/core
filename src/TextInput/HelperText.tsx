import styled, { css } from 'styled-components';

export interface HelperTextProps {
  disabeld?: boolean;
  error?: boolean;
  variant?: 'standard' | 'filled' | 'outlined';
  dense?: boolean;
}

const containedStyle = css`
  margin: 8px 12px 0;
`;

/**
 * 帮助性文本
 */
const HelperText = styled.p<HelperTextProps>`
  color: ${(props) => props.theme.palette.text.secondary};
  font-family: ${(props) => props.theme.typography.fontFamily};
  font-size: ${(props) => props.theme.typography.caption.fontSize};
  text-align: left;
  line-height: 0.75rem;
  min-height: 0.75rem;
  margin: 0;
  margin-top: ${(props) => (props.dense ? 4 : 8)}px;

  ${(props) => props.disabeld && `color:${props.theme.palette.text.disabled}`};
  ${(props) => props.error && `color:${props.theme.palette.error.main}`};
  ${(props) =>
    (props.variant === 'filled' || props.variant === 'outlined') &&
    containedStyle};
`;

export default HelperText;
