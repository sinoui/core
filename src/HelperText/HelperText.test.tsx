import { render, cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import renderer from 'react-test-renderer';
import React from 'react';
import 'jest-styled-components';
import HelperText from './HelperText';

afterEach(cleanup);

it('渲染帮助性文本', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <HelperText data-testid="helperText">帮助性文本</HelperText>
    </ThemeProvider>,
  );

  const helperText = getByTestId('helperText');

  expect(helperText).toHaveStyleRule('font-size', '0.75rem');
  expect(helperText).toHaveStyleRule(
    'color',
    defaultTheme.palette.text.secondary.replace(/ +/g, ''),
  );
});

it('渲染验证错误信息', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <HelperText data-testid="helperText" error>
        必填
      </HelperText>
    </ThemeProvider>,
  );

  const helperText = getByTestId('helperText');

  expect(helperText).toHaveStyleRule(
    'color',
    defaultTheme.palette.error.main.replace(/ +/g, ''),
  );
});

it('不可用情况', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <HelperText data-testid="helperText" disabled>
        必填
      </HelperText>
    </ThemeProvider>,
  );

  const helperText = getByTestId('helperText');

  expect(helperText).toHaveStyleRule(
    'color',
    defaultTheme.palette.text.disabled.replace(/ +/g, ''),
  );
});

it('使用不同的颜色', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <HelperText data-testid="helperText" color="primary">
        必填
      </HelperText>
    </ThemeProvider>,
  );

  const helperText = getByTestId('helperText');

  expect(helperText).toHaveStyleRule(
    'color',
    defaultTheme.palette.primary.main.replace(/ +/g, ''),
  );
});

it('快照测试', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <>
          <HelperText>帮助性文本</HelperText>
          <HelperText error>错误信息</HelperText>
          <HelperText disabled>不可用</HelperText>
          <HelperText color="primary">指定其他颜色</HelperText>
        </>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
