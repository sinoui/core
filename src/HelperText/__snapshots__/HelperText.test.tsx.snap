// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 1`] = `
Array [
  .c0 {
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  -webkit-letter-spacing: 0.03333em;
  -moz-letter-spacing: 0.03333em;
  -ms-letter-spacing: 0.03333em;
  letter-spacing: 0.03333em;
  color: rgba(0,0,0,0.6);
  text-align: left;
}

<div
    className="c0 sinoui-helper-text"
  >
    帮助性文本
  </div>,
  .c0 {
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  -webkit-letter-spacing: 0.03333em;
  -moz-letter-spacing: 0.03333em;
  -ms-letter-spacing: 0.03333em;
  letter-spacing: 0.03333em;
  color: #f44336;
  text-align: left;
}

<div
    className="c0 sinoui-helper-text"
  >
    错误信息
  </div>,
  .c0 {
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  -webkit-letter-spacing: 0.03333em;
  -moz-letter-spacing: 0.03333em;
  -ms-letter-spacing: 0.03333em;
  letter-spacing: 0.03333em;
  color: rgba(0,0,0,0.38);
  text-align: left;
}

<div
    className="c0 sinoui-helper-text"
    disabled={true}
  >
    不可用
  </div>,
  .c0 {
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  -webkit-letter-spacing: 0.03333em;
  -moz-letter-spacing: 0.03333em;
  -ms-letter-spacing: 0.03333em;
  letter-spacing: 0.03333em;
  color: #3f51b5;
  text-align: left;
}

<div
    className="c0 sinoui-helper-text"
    color="primary"
  >
    指定其他颜色
  </div>,
]
`;
